mantidproject/mantid
# ReplicateMD checks of input dimensions

Modified the validation method for the input dimensions to allow any number of "trailing" integrated dimensions. Added an error message if a particular condition isn't met.

**To test:**
1. Run the following script. It should fail on the last executable statement. Check that the warning log message makes sense.
2. Uncomment the last 3 statements and run them only. They should run without error and created `replicated` workspace must be filled in correctly.

```
#Some problems with MD algorithms / workflow

#Create an MD workspace, file backed because my system might not have enough memory:

import os
import numpy as np
from os import path
this_dir = path.dirname(__file__)

#Create projections workspace for later cutting
from mantid.api import Projection
# Create an identity projection
proj_id = Projection([1,1,0], [-1,1,0], [0,0,1])
proj_id.setType(0, 'r') 
proj_id.setType(1, 'r')
proj_id.setType(2, 'r') 
ws_id = proj_id.createWorkspace()

#Make MD mega file (like gen_sqw / accumulate_sqw)
psi_arr=np.arange(0.0,92.0,2)
run_arr=range(15052,15098)

input_runs = list()
psi = list()
gs = list()
gl = list()
#note that gl, gs etc are not mandtory inputs. If we do not specify them then they are set to zero.

input_runs = list()
psi = list()
gs = list()
gl = list()
for i in range(2):
    source_file = ('%s/map'+str(run_arr[i])+'_ei400.nxspe') % this_dir
    input_runs.append(source_file)
    psi.append(float(psi_arr[i]))
    gl.append(0.0)
    gs.append(0.0)

print(psi)
print(input_runs)

md_ws = CreateMD(input_runs, Emode='Direct', Alatt=[2.87,2.87,2.87], Angdeg=[ 90, 90, 90], u=[1, 0, 0,], v=[0,1,0], Psi=psi, Gl=gl, Gs=gs, EFix=400.0,Filename='test.nxs',FileBackEnd=True)
#Up to this point everything works as we would expect it to.

########################################

##Problem with ReplicateMD logic:

#2D slice (the shape workspace)
out_md_2d = CutMD(md_ws, Projection=ws_id, PBins=([-3,0.05,3],[-1.1,-0.9],[-0.1,0.1],[0,4,360]), NoPix=True)

#1D cut (the data workspace)
out_md_1d_nopix= CutMD(md_ws, Projection=ws_id, PBins=([2,3],[-1.1,-0.9],[-0.1,0.1],[0,4,360]), NoPix=True)

#Attempt to replicate data to shape:
replicated = ReplicateMD(out_md_2d, out_md_1d_nopix)

# Uncomment and run these lines:
#TransposeMD(out_md_2d, [0, 3, 1, 2], OutputWorkspace='t2d')
#TransposeMD(out_md_1d_nopix, [0, 3, 1, 2], OutputWorkspace='t1d')
#replicated = ReplicateMD('t2d', 't1d')
```

Fixes #16389.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
